src/transformStateWithClones.js

```diff
@@ -7,7 +7,36 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  let newState = { ...state };
+  const arrOfCopies = [];
+
+  const ADD = 'addProperties';
+  const REMOVE = 'removeProperties';
+  const CLEAR = 'clear';
+
+  for (const action of actions) {
+    let cloneState = { ...newState };
+
+    switch (action.type) {
+      case ADD:
+        Object.assign(cloneState, action.extraData);
+        break;
+
+      case REMOVE:
+        for (const key of action.keysToRemove) {
+          delete cloneState[key];
+        }
+        break;
+
+      case CLEAR:
+        cloneState = {};
+        break;
+    }
+    arrOfCopies.push(cloneState);
+    newState = cloneState;
+  }
```

Comment:
would be a bit better, but you solution also good
```suggestion
  for (const action of actions) {
    switch (action.type) {
      case ADD:
        Object.assign(newState, action.extraData);
        break;

      case REMOVE:
        for (const key of action.keysToRemove) {
          delete newState[key];
        }
        break;

      case CLEAR:
        newState= {};
        break;
    }
    
    arrOfCopies.push({...newState });
  }
```